### PR TITLE
spotify-adblock: init at 1.4

### PR DESCRIPTION
**Pull request by milahu**
###### Motivation for this change

remove ads from "spotify free"

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/audio/spotify/spotify-adblock.nix [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+    mkdir -p $out/{lib,bin}
+
+    cp spotify-adblock.so $out/lib/
+
+    # wrapper
+    str=""
+    str+="#!${pkgs.runtimeShell}\n"
+    str+="LD_PRELOAD=$out/lib/spotify-adblock.so exec ${pkgs.spotify}/bin/spotify\n"
+    echo "$str" >$out/bin/spotify-adblock
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/spotify/spotify-adblock.nix:78]:
```suggestion
    echo "#!${pkgs.runtimeShell}
    LD_PRELOAD=$out/lib/spotify-adblock.so exec ${pkgs.spotify}/bin/spotify
    " >$out/bin/spotify-adblock
```

File: pkgs/applications/audio/spotify/spotify-adblock.nix [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+  buildInputs = [
+    pkgs.spotify
+  ];
+
+  preBuild = ''
+    ln -s -v "${cef-src}/include" ./
+
+    # debug
+    echo ls include/capi/
+    ls include/capi/
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/spotify/spotify-adblock.nix:58]:
```suggestion
```

File: pkgs/applications/audio/spotify/spotify-adblock.nix [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+*/
+
+{ pkgs, lib, stdenv, fetchFromGitHub, fetchFromBitbucket }:
+
+stdenv.mkDerivation rec {
+
+  pname = "spotify-adblock";
+  version = "1.4";
+  name = "${pname}-${version}";
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/spotify/spotify-adblock.nix:25]:
```suggestion
stdenv.mkDerivation rec {
  pname = "spotify-adblock";
  version = "1.4";
```

File: pkgs/applications/audio/spotify/spotify-adblock.nix [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+  */
+
+  # this should be enough for build:
+  #git clone --depth 1  https://bitbucket.org/chromiumembedded/cef.git --branch 3729
+  # -> 30 MByte
+
+  meta = with lib; {
+    description = "Spotify adblocker for Linux";
+    homepage = "https://github.com/abba23/spotify-adblock-linux";
+    license = licenses.gpl3;
```
[the remark this answers is not in the document]
Comment (milahu) [reply pkgs/applications/audio/spotify/spotify-adblock.nix]: only

**Comment by milahu** [comment]
took a while to get this working ... 

problem was the old spotify version in nixpkgs unstable (black screen bug)
tried to downgrade but that caused only more problems (glibc version mismatch)

here is a `/etc/nixos/configuration.nix` to get this working with the [spotify 1.1.55.498](https://github.com/NixOS/nixpkgs/pull/118055)

```nix
{ config, pkgs, lib, stdenv, ... }:
let
  base-config = {
    allowUnfree = true; # needed for spotify, ...
  };
  spotify-pkgs = import (fetchTarball
    https://github.com/NixOS/nixpkgs/archive/caeafab22e39f1e95fc9452ff190fa7bdf342579.tar.gz # 1.1.55.498
    #https://github.com/NixOS/nixpkgs/archive/bed08131cd29a85f19716d9351940bdc34834492.tar.gz # 1.1.46.916 2021-02-08 # black screen bug
    #https://github.com/NixOS/nixpkgs/archive/2c162d49cd5b979eb66ff1653aecaeaa01690fcc.tar.gz # 1.1.26.501 2020-11-09 # glibc version mismatch
    #https://github.com/NixOS/nixpkgs/archive/fcc8660d359d2c582b0b148739a72cec476cfef5.tar.gz # 1.1.10.546 2020-03-09
  ) { config = base-config; };
in
{
  imports = [ ./hardware-configuration.nix ];
  nixpkgs.config = base-config // {
    packageOverrides = pkgs: with pkgs; {
      inherit (spotify-pkgs) spotify;
      inherit (spotify-pkgs) spotify-unwrapped; # required by spotify-adblock
    };
  };
  nixpkgs.overlays = [ ( self: super: {
    spotify-adblock = pkgs.callPackage ./extra-pkgs/spotify/spotify-adblock.nix {};
  })];
  environment.systemPackages = with pkgs; [
    spotify-adblock
  ];
}
```

```
sudo mkdir -p /etc/nixos/extra-pkgs/spotify

sudo wget -O /etc/nixos/extra-pkgs/spotify/spotify-adblock.nix \
https://raw.githubusercontent.com/milahu/nixpkgs/patch-6/pkgs/applications/audio/spotify/spotify-adblock.nix

sudo nixos-rebuild switch

spotify-adblock
```

File: pkgs/applications/audio/spotify/spotify-adblock.nix [hunk 5]
```diff
@@ -0,0 +1,217 @@
+{ fetchFromGitHub, spotify, spotify-unwrapped,
+  gtk3, libxkbcommon, # spotify-unwrapped-1.1.55.498
+  # copy paste from spotify/default.nix:
+  fetchurl, lib, stdenv, squashfsTools, xorg, alsaLib, makeWrapper, openssl, freetype
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/spotify/spotify-adblock.nix:4]:
```suggestion
{ fetchFromGitHub, spotify, spotify-unwrapped
, gtk3, libxkbcommon # spotify-unwrapped-1.1.55.498
  # copy paste from spotify/default.nix:
, fetchurl, lib, stdenv, squashfsTools, xorg, alsaLib, makeWrapper, openssl, freetype
```

File: pkgs/applications/audio/spotify/spotify-adblock.nix [hunk 6]
```diff
@@ -1,29 +1,156 @@
...
+let
+  curlWithGnutls = curl.override { gnutlsSupport = true; sslSupport = false; }; # provides libcurl-gnutls.so.4
+  # warning message: libcurl-gnutls.so.4: no version information available
+  # spotify seems to want curl_3 = libcurl 7.15.x = older than 2006-10
+  # readelf -s /nix/store/*-spotify-unwrapped-*/share/spotify/.spotify-wrapped | grep -i curl
+  # 520: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND [...]@CURL_GNUTLS_3 (19)
+
+  # copy paste from spotify/default.nix:
+  deps =
+    [
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/spotify/spotify-adblock.nix:18]:
```suggestion
  deps = [
```

**Review by SuperSandro2000 — COMMENTED** [review]
I am not sure if nixpkgs is the right place for this.

**Comment by milahu** [comment]
> I am not sure if nixpkgs is the right place for this.

agree ... also in arch linux, adblockers like `blockify` or `spotify-adblock-linux` are in the [AUR](https://aur.archlinux.org/packages/?K=spotify)

File: pkgs/applications/audio/spotify/default.nix [hunk 7]
```diff
@@ -63,6 +63,9 @@ let
     zlib
   ];
 
+  callPackage = newScope spotify;
+  spotify-adblock = callPackage ./spotify-adblock.nix {};
+
```
Comment (milahu) [inline pkgs/applications/audio/spotify/default.nix:68]: broken. `spotify` is not defined here

**Comment by milahu** [comment]
closing in favor of 

> published to NUR as <code>[nur.repos.milahu](https://nur.nix-community.org/repos/milahu/).spotify-adblock-linux</code>

**Pull request closed without merging**